File: app/src/main/java/com/chisto/Activities/OrdersActivity.java
package com.chisto.Activities;

import com.google.gson.JsonObject;

import android.app.ProgressDialog;
import android.content.Intent;
import android.os.Bundle;
import android.os.Handler;
import android.support.annotation.Nullable;
import android.support.v4.content.ContextCompat;
import android.support.v7.widget.DefaultItemAnimator;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.helper.ItemTouchHelper;
import android.view.View;
import android.widget.ImageView;
import android.widget.TextView;

import com.chisto.Adapters.OrdersAdapter;
import com.chisto.Base.BaseActivity;
import com.chisto.Custom.RecyclerListView;
import com.chisto.R;
import com.chisto.Server.ServerApi;
import com.chisto.Server.ServerConfig;
import com.chisto.Utils.Animations.Animations;
import com.chisto.Utils.Image;
import com.chisto.Utils.OrderList;
import com.chisto.Utils.SwipeItemDecoration;
import com.chisto.Utils.TouchHelper;
import com.crashlytics.android.Crashlytics;

import io.fabric.sdk.android.Fabric;
import retrofit2.Call;
import retrofit2.Callback;
import retrofit2.Response;

public class OrdersActivity extends BaseActivity {

    private OrdersAdapter adapter;

    private TextView contBtn;
    private static boolean dialogOpened = false;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        Fabric.with(this, new Crashlytics());
        setContentView(R.layout.activity_orders);

        contBtn = (TextView) findViewById(R.id.textView2);

        initRecyclerView();
        setOnClickListeners();
    }

    private void initRecyclerView() {
        RecyclerListView view = (RecyclerListView) findViewById(R.id.recyclerView);
        view.setLayoutManager(new LinearLayoutManager(this));
        view.setItemAnimator(new DefaultItemAnimator());
        view.setHasFixedSize(true);
        view.setEmptyView(findViewById(R.id.empty_orders));

        adapter = new OrdersAdapter(this);
        view.setAdapter(adapter);

        ItemTouchHelper mItemTouchHelper = new ItemTouchHelper(new TouchHelper(0, ItemTouchHelper.LEFT, this, view));
        mItemTouchHelper.attachToRecyclerView(view);
        view.addItemDecoration(new SwipeItemDecoration());
    }

    private void setOnClickListeners() {
        findViewById(R.id.left_btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(OrdersActivity.this, ProfileActivity.class);
                startActivity(intent);
            }
        });

        findViewById(R.id.right_btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(OrdersActivity.this, SelectCategoryActivity.class);
                startActivity(intent);
            }
        });

        findViewById(R.id.add_btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(OrdersActivity.this, SelectCategoryActivity.class);
                startActivity(intent);
            }
        });

        findViewById(R.id.dialog).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                new Handler().post(new Runnable() {
                    @Override
                    public void run() {
                        dialogOpened = false;
                        Animations.animateRevealHide(findViewById(R.id.dialog));
                    }
                });
            }
        });

        findViewById(R.id.main_dialog_part).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                // TODO
            }
        });
    }

    @Override
    public void onBackPressed() {
        super.onBackPressed();
        if (dialogOpened) {
            new Handler().post(new Runnable() {
                @Override
                public void run() {
                    dialogOpened = false;
                    Animations.animateRevealHide(findViewById(R.id.dialog));
                }
            });
        }
    }

    @Override
    protected void onResume() {
        super.onResume();
        update();
    }

    @SuppressWarnings("ConstantConditions")
    private void update() {
        if (OrderList.get() != null) {
            adapter.setCollection(OrderList.get());
            adapter.notifyDataSetChanged();

            if (adapter.getItemCount() != 0) {
                contBtn.setText(R.string.cont_code_str);
                contBtn.setBackgroundColor(ContextCompat.getColor(OrdersActivity.this, R.color.colorPrimary));
                contBtn.setOnClickListener(new View.OnClickListener() {
                    @Override
                    public void onClick(View view) {
                        loadLastOrder();
                    }
                });
            } else {
                contBtn.setText(R.string.nothing_selected_code_str);
                contBtn.setBackgroundColor(ContextCompat.getColor(OrdersActivity.this, R.color.greyColor));
                contBtn.setOnClickListener(null);
            }
        }
    }

    private void loadLastOrder() {
        final ProgressDialog dialog = new ProgressDialog(this);
        dialog.show();

        ServerApi.get(OrdersActivity.this).api().getLaundry(1).enqueue(new Callback<JsonObject>() {
            @Override
            public void onResponse(Call<JsonObject> call, Response<JsonObject> response) {
                dialog.dismiss();
                if (response.isSuccessful()) {
                    parseAnswer(response.body());
                    new Handler().post(new Runnable() {
                        @Override
                        public void run() {
                            dialogOpened = true;
                            Animations.animateRevealShow(findViewById(R.id.dialog), OrdersActivity.this);
                        }
                    });
                } else {
                    onInternetConnectionError();
                }
            }

            @Override
            public void onFailure(Call<JsonObject> call, Throwable t) {
                dialog.dismiss();
                onInternetConnectionError();
            }
        });
    }

    private void parseAnswer(JsonObject object) {
        ((TextView) findViewById(R.id.name_text)).setText(getString(R.string.laundary_code_str) + object.get("name").getAsString());
        ((TextView) findViewById(R.id.desc_text)).setText(object.get("description").getAsString());
        ((TextView) findViewById(R.id.order_current_btn)).setText(getString(R.string.order_code_str) + object.get("name").getAsString());

        Image.loadPhoto(ServerConfig.INSTANCE.getImageUrl() +
                object.get("background_image_url").getAsString(), (ImageView) findViewById(R.id.back_image));
        Image.loadPhoto(ServerConfig.INSTANCE.getImageUrl() +
                object.get("logo_url").getAsString(), (ImageView) findViewById(R.id.logo_image));

        findViewById(R.id.cont_btn).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View view) {
                Intent intent = new Intent(OrdersActivity.this, LaundriesActivity.class);
                startActivity(intent);
            }
        });
    }
}

File: app/src/main/java/ru/binaryblitz/Chisto/ui/order/OrdersActivity.kt
package ru.binaryblitz.Chisto.ui.order

import android.app.Activity
import android.app.ProgressDialog
import android.content.Intent
import android.os.Bundle
import android.os.Handler
import android.support.v7.widget.DefaultItemAnimator
import android.support.v7.widget.LinearLayoutManager
import android.view.Menu
import android.view.MenuItem
import android.view.View
import android.widget.EditText
import android.widget.TextView
import com.afollestad.materialdialogs.MaterialDialog
import com.crashlytics.android.Crashlytics
import com.google.gson.JsonObject
import com.iarcuschin.simpleratingbar.SimpleRatingBar
import io.fabric.sdk.android.Fabric
import kotlinx.android.synthetic.main.activity_orders.*
import kotlinx.android.synthetic.main.dialog_review.*
import kotlinx.android.synthetic.main.fragment_empty_order_list.*
import kotlinx.android.synthetic.main.popup_your_order.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import ru.binaryblitz.Chisto.R
import ru.binaryblitz.Chisto.R.string
import ru.binaryblitz.Chisto.entities.MyOrder
import ru.binaryblitz.Chisto.network.DeviceInfoStore
import ru.binaryblitz.Chisto.network.ServerApi
import ru.binaryblitz.Chisto.ui.base.BaseActivity
import ru.binaryblitz.Chisto.ui.categories.CategoryActivity
import ru.binaryblitz.Chisto.ui.laundries.LaundriesActivity
import ru.binaryblitz.Chisto.ui.order.ItemInfoActivity.EXTRA_COLOR
import ru.binaryblitz.Chisto.ui.order.adapters.OrdersAdapter
import ru.binaryblitz.Chisto.utils.AndroidUtilities
import ru.binaryblitz.Chisto.utils.Animations
import ru.binaryblitz.Chisto.utils.OrderList
import ru.binaryblitz.Chisto.views.RecyclerListView

class OrdersActivity : BaseActivity() {

    private lateinit var adapter: OrdersAdapter
    private lateinit var removeBtn: MenuItem
    private var dialogOpened = false

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        Fabric.with(this, Crashlytics())
        setContentView(R.layout.activity_orders)

        initViewElements()
        setOnClickListeners()

        Handler().post {
            if (newOrderId != 0) {
                showOrderDialog(newOrderId)
            } else {
                getUser()
            }
        }
    }

    private fun getUser() {
        ServerApi.get(this).api().getUser(DeviceInfoStore.getToken(this)).enqueue(object : Callback<JsonObject> {
            override fun onResponse(call: Call<JsonObject>, response: Response<JsonObject>) {
                if (response.isSuccessful) {
                    parseAnswer(response.body()!!)
                }
            }

            override fun onFailure(call: Call<JsonObject>, t: Throwable) {
            }
        })
    }

    private fun showOrderDialog(id: Int) {
        Handler().post {
            dialogOpened = true
            order_name.text = getString(R.string.number_sign) + id.toString()
            Animations.animateRevealShow(findViewById(R.id.dialog_new_order), this@OrdersActivity)
            newOrderId = 0
        }
    }

    private fun parseAnswer(obj: JsonObject) {
        val order = obj.get("order")

        if (order == null || obj.get("order").isJsonNull) {
            return
        }

        val myOrder = MyOrder(order.asJsonObject)

        if (myOrder.status != MyOrder.Status.COMPLETED) {
            return
        }

        val review = order.asJsonObject.get("rating") ?: return

        laundryId = AndroidUtilities.getIntFieldFromJson(order.asJsonObject.get("laundry").asJsonObject.get("id"))

        if (review.isJsonNull) {
            showReviewDialog(AndroidUtilities.getIntFieldFromJson(order.asJsonObject.get("id")))
        }
    }

    private fun sendReview() {
        val dialog = ProgressDialog(this)
        dialog.show()
        ServerApi.get(this).api().sendReview(laundryId, generateJson(), DeviceInfoStore.getToken(this)).enqueue(object : Callback<JsonObject> {
            override fun onResponse(call: Call<JsonObject>?, response: Response<JsonObject>) {
                dialog.dismiss()
                Animations.animateRevealHide(findViewById(R.id.dialog))
                if (response.isSuccessful) {
                    parseReviewResponse()
                } else {
                    onServerError(response)
                }
            }

            override fun onFailure(call: Call<JsonObject>?, t: Throwable?) {
                dialog.dismiss()
                Animations.animateRevealHide(findViewById(R.id.dialog))
                onInternetConnectionError()
            }
        })
    }

    private fun showErrorDialog() {
        MaterialDialog.Builder(this)
                .title(getString(R.string.error))
                .content(getString(R.string.wrong_review_code))
                .positiveText(R.string.ok_code)
                .onPositive { dialog, _ -> dialog.dismiss() }
                .show()
    }

    private fun checkReview(): Boolean {
        return (findViewById<SimpleRatingBar>(R.id.ratingBar)).rating.toInt() != 0
    }

    private fun generateJson(): JsonObject {
        val obj = JsonObject()

        obj.addProperty("value", (findViewById<SimpleRatingBar>(R.id.ratingBar)).rating)
        obj.addProperty("content", (findViewById<EditText>(R.id.review_text)).text.toString())

        val toSend = JsonObject()
        toSend.add("rating", obj)

        return toSend
    }

    private fun parseReviewResponse() {
        Animations.animateRevealHide(findViewById(R.id.dialog))
    }

    private fun showReviewDialog(id: Int) {
        Handler().post {
            dialogOpened = true
            (findViewById<TextView>(R.id.order_name_completed)).text =
                    getString(R.string.order) + " № " + id.toString() + getString(R.string.completed)
            Animations.animateRevealShow(findViewById(R.id.dialog), this@OrdersActivity)
        }
    }

    private fun initViewElements() {
        setSupportActionBar(toolbar)
        val view = findViewById<RecyclerListView>(R.id.recyclerView)
        view.layoutManager = LinearLayoutManager(this)
        view.itemAnimator = DefaultItemAnimator()
        view.setHasFixedSize(true)
        view.emptyView = findViewById(R.id.empty_orders)

        adapter = OrdersAdapter(this)
        adapter.setItemColor(intent.getStringExtra(EXTRA_COLOR))
        view.adapter = adapter
        adapter.onItemSelectAction.subscribe { isSelected -> showSelection(isSelected) }
    }

    private fun showSelection(selected: Boolean) {
        setToolbarTitle(selected)
        removeBtn.isVisible = selected
        update()
    }

    private fun setToolbarTitle(selected: Boolean) {
        dateTextView.text = if (selected) {
            getString(string.select_order)
        } else {
            getString(string.order)
        }
    }

    private fun openActivity(activity: Class<out Activity>) {
        val intent = Intent(this@OrdersActivity, activity)
        startActivity(intent)
    }

    private fun setOnClickListeners() {
        left_btn.setOnClickListener { finish() }
        add_btn.setOnClickListener { openActivity(CategoryActivity::class.java) }
        cont_btn.setOnClickListener {
            if (!checkReview()) {
                showErrorDialog()
            } else {
                sendReview()
            }
        }
        new_order_dialog_btn.setOnClickListener {
            Animations.animateRevealHide(findViewById(R.id.dialog_new_order))
            getUser()
        }

        addItemButton.setOnClickListener { openActivity(CategoryActivity::class.java) }
    }

    override fun onBackPressed() {
        if (dialogOpened) {
            return
        }
        super.onBackPressed()
    }

    override fun onResume() {
        super.onResume()
        update()
    }

    private fun update() {
        if (OrderList.get() == null) {
            return
        }

        adapter.setCollection(OrderList.get()!!)

        if (adapter.itemCount != 0) {
            addItemButton.visibility = View.VISIBLE
            setContinueButtonEnabled()
        } else {
            addItemButton.visibility = View.GONE
            setContinueButtonDisabled()
        }
    }

    private fun setContinueButtonEnabled() {
        continueButton.setText(R.string.continue_btn)
        continueButton.setOnClickListener {
            LaundriesActivity.longTreatment = adapter.hasItemsWithLongTreatment()
            val intent = Intent(this@OrdersActivity, LaundriesActivity::class.java)
            startActivity(intent)
        }
    }

    private fun setContinueButtonDisabled() {
        continueButton.setText(R.string.nothing_selected)
        continueButton.isEnabled = false
        continueButton.setOnClickListener(null)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        menuInflater.inflate(R.menu.menu_order, menu)
        removeBtn = menu.findItem(R.id.action_remove)
        removeBtn.isVisible = false
        return true
    }

    override fun onOptionsItemSelected(item: MenuItem?): Boolean {
        when (item?.itemId) {
            R.id.action_remove -> {
                adapter.removeSelectedOrders()
            }
        }
        return super.onOptionsItemSelected(item)
    }

    companion object {
        var laundryId = 0
        var newOrderId = 0
    }
}
